src/backend/commands/dirtablecmds.c

```diff
@@ -0,0 +1,472 @@
+
+		MemoryContextSwitchTo(oldContext);
+	}
+
+	funcCtx = SRF_PERCALL_SETUP();
+	context = (TableFunctionContext *) funcCtx->user_fctx;
+
+	if (table_scan_getnextslot(context->scanDesc, ForwardScanDirection, context->slot))
+	{
+		Datum attrRelativePath;
+		Datum attrSize;
+		Datum attrLastModified;
+		Datum attrMd5Sum;
+		Datum attrTags;
+		bool  isNull;
+		char  *scopedFileUrl;
+
+		slot_getallattrs(context->slot);
+
+		attrRelativePath  = slot_getattr(context->slot, 1, &isNull);
+		Assert(isNull == false);
+		attrSize = slot_getattr(context->slot, 2, &isNull);
+		Assert(isNull == false);
+		attrLastModified = slot_getattr(context->slot, 3, &isNull);
+		Assert(isNull == false);
+		attrMd5Sum = slot_getattr(context->slot, 4, &isNull);
+		Assert(isNull == false);
+		attrTags = slot_getattr(context->slot, 5, &isNull);
+
+		MemSet(values, 0, sizeof(values));
+		MemSet(nulls, 0, sizeof(nulls));
+
+		scopedFileUrl = getScopedFileUrl(context->dirTable, TextDatumGetCString(attrRelativePath));
+
+		values[0] = PointerGetDatum(cstring_to_text(scopedFileUrl));
+		values[1] = attrRelativePath;
+		values[2] = attrTags;
+		nulls[2] = isNull;
+		values[3] = attrSize;
+		values[4] = attrLastModified;
+		values[5] = attrMd5Sum;
+		values[6] = getFileContent(context->dirTable->spcId, scopedFileUrl);
+
+		tuple = heap_form_tuple(funcCtx->tuple_desc, values, nulls);
+		result = HeapTupleGetDatum(tuple);
+
+		SRF_RETURN_NEXT(funcCtx, result);
+	}
+
+	table_endscan(context->scanDesc);
+	ExecDropSingleTupleTableSlot(context->slot);
+	table_close(context->relation, AccessShareLock);
+
+	pfree(context);
+	funcCtx->user_fctx = NULL;
+
+	SRF_RETURN_DONE(funcCtx);
+}
+
+Datum
+remove_file_segment(PG_FUNCTION_ARGS)
+{
+	Relation 	relation;
+	ScanKeyData scankey;
+	SysScanDesc scan;
+	HeapTuple 	tuple;
+	Oid 		indexOid;
+	List		*indexOids;
+	Oid 		relId;
+	char 		*relativePath;
+	char 		*fullPathName;
+	bool 		exist = false;
+	DirectoryTable	*dirTable;
+
+	if (PG_ARGISNULL(0))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("relation name cannot be NULL")));
+	relId = PG_GETARG_OID(0);
+
+	if (PG_ARGISNULL(1))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("file path cannot be null")));
+	relativePath = PG_GETARG_CSTRING(1);
+
+	dirTable = GetDirectoryTable(relId);
+	relation = table_open(relId, AccessExclusiveLock);
+
+	SIMPLE_FAULT_INJECTOR("remove_file_inject");
+
+	indexOids = RelationGetIndexList(relation);
+	Assert(list_length(indexOids) == 1);
+
+	indexOid = list_nth_oid(indexOids, 0);
+	ScanKeyInit(&scankey,
+			 	1, /* relative_path */
+			 	BTEqualStrategyNumber, F_TEXTEQ,
+			 	CStringGetTextDatum(relativePath));
+
+	scan = systable_beginscan(relation, indexOid, true, NULL, 1, &scankey);
+	tuple = systable_getnext(scan);
+	if (HeapTupleIsValid(tuple))
+	{
+		CatalogTupleDelete(relation, &tuple->t_self);
+		fullPathName = psprintf("%s/%s", dirTable->location, relativePath);
```

Comment:
suppose `dirTable->location` assign to `/usr/local/lib`, then `relativePath ` equal to `libgcc_s.so.1 `  , will delete local system file ?

Reply:
Actually not, as `/usr/local/lib` has root privileges while we connect to db as another user which is not permitted.